compaction/compaction_manager.cc

```diff
@@ -756,15 +758,21 @@ future<> compaction_manager::postponed_compactions_reevaluation() {
             _postponed.clear();
             co_return;
         }
-        auto postponed = std::move(_postponed);
+        auto postponed = std::exchange(_postponed, {});
         try {
-            for (auto& t : postponed) {
+            auto holders = boost::copy_range<std::vector<seastar::gate::holder>>(postponed | boost::adaptors::transformed([this] (compaction::table_state* t) {
+                return get_compaction_state(t).gate.hold();
```

Comment:
If this throws, shouldn't we erase it from `postponed`?
Otherwise the catch clause will return it to `_postponed` and we'll never get rid of it.

Reply:
this should never throw. as all table_state in _postponed must have a compaction_state registered for it.

note that compaction_manager::remove() is now erasing table_state from _postponed after it closed and waited for gate.  _compact_state only has table_state removed from it after we have removed the table_state from _postponed.

Reply:
It seems like this ping-pong between the 2 functions is overly complicated and fragile.
If remove makes sure to erase thextable_state feom _postponed, maybe reevaluate shpuld loop while (!_postponed.empty) and just dequeue and submit the back() entry, one at a time, between maybe_yield.  submit() chcks for can_proceed so it's suffcient that the table is valid at the time submit is called.

Reply:
this way there is no need to exchange the list and insert the remainder from thectemporary vector back into the member

Reply:
there's really no need to play all these tricky games to recover from exception.

Reply:
Why do we need to grab _postponed into a local varuable? why not grab the table states one at a time?, as it's much simpler?

Reply:
I'd need to ask Raphael from ~4 years old. I don't what was in his mind :-) There's no need, indeed.

Reply:
I'd need to ask Raphael from ~4 years ago. I don't what was in his mind :-) There's no need, indeed.

Reply:
:-)

Reply:
Ok, I think the reason was, maybe, to avoid reentrancy, as submit() itself could end up postponing new tasks. therefore you could run into a very long loop. we can easily avoid it by deferring erase() to after submit() was called.

Reply:
Ok, that's it. A postponed task_state which gets postponed again remains in postponed state. In simple words, a task_state that attemps another compaction should be given the ability to reinsert itself into postponed list. That's why I decided to move _postponed into local.

Reply:
Yes, that makes sense.
Let's document that in a comment please, while at it.

Reply:
I wish I could remove this move into local. 

Perhaps do this:

```
diff --git a/compaction/compaction_manager.cc b/compaction/compaction_manager.cc
index 681fa2c7d2..69341a9ae4 100644
--- a/compaction/compaction_manager.cc
+++ b/compaction/compaction_manager.cc
@@ -15,6 +15,7 @@
 #include <seastar/core/coroutine.hh>
 #include <seastar/coroutine/switch_to.hh>
 #include <seastar/coroutine/parallel_for_each.hh>
+#include <seastar/coroutine/maybe_yield.hh>
 #include "sstables/exceptions.hh"
 #include "sstables/sstable_directory.hh"
 #include "locator/abstract_replication_strategy.hh"
@@ -744,8 +745,9 @@ void compaction_manager::enable() {
 std::function<void()> compaction_manager::compaction_submission_callback() {
     return [this] () mutable {
         for (auto& e: _compaction_state) {
-            submit(*e.first);
+            postpone_compaction_for_table(e.first);
         }
+        reevaluate_postponed_compactions();
     };
 }
 
@@ -756,15 +758,24 @@ future<> compaction_manager::postponed_compactions_reevaluation() {
             _postponed.clear();
             co_return;
         }
-        auto postponed = std::move(_postponed);
         try {
-            for (auto& t : postponed) {
+            std::unordered_set<compaction::table_state*> processed;
+            while (!_postponed.empty()) {
+                compaction::table_state* t = *_postponed.begin();
+                // A postponed task_state may re-insert itself into postponed list. But we don't need to process
+                // it twice in the same reevaluation iteration.
+                if (!processed.insert(t).second) {
+                    continue;
+                }
+                _postponed.erase(t);
                 auto s = t->schema();
                 cmlog.debug("resubmitting postponed compaction for table {}.{} [{}]", s->ks_name(), s->cf_name(), fmt::ptr(t));
                 submit(*t);
+                co_await coroutine::maybe_yield();
             }
         } catch (...) {
-            _postponed = std::move(postponed);
+            cmlog.warn("Compaction manager failed when reevaluating postponed compaction(s), due to {}", std::current_exception());
+            // swallow exception to allow next iteration.
         }
     }
 }

```

I am not very happy about it, but maybe it's better than moving postponed into local and having complicated scheme for remove() to not miss task_state stored in local postponed list.

What do you think?

Reply:
we can iterate through all table_state twice now, but factor of 2 is a low-order factor, therefore complexity remains O(N).

Reply:
Actually that can produce terrible complexity. Taking my word back.

Reply:
I think it can be simplified as follows:
```diff
diff --git a/compaction/compaction_manager.cc b/compaction/compaction_manager.cc
index 681fa2c7d2..eae931421b 100644
--- a/compaction/compaction_manager.cc
+++ b/compaction/compaction_manager.cc
@@ -15,6 +15,7 @@
 #include <seastar/core/coroutine.hh>
 #include <seastar/coroutine/switch_to.hh>
 #include <seastar/coroutine/parallel_for_each.hh>
+#include <seastar/coroutine/maybe_yield.hh>
 #include "sstables/exceptions.hh"
 #include "sstables/sstable_directory.hh"
 #include "locator/abstract_replication_strategy.hh"
@@ -744,8 +745,9 @@ void compaction_manager::enable() {
 std::function<void()> compaction_manager::compaction_submission_callback() {
     return [this] () mutable {
         for (auto& e: _compaction_state) {
-            submit(*e.first);
+            postpone_compaction_for_table(e.first);
         }
+        reevaluate_postponed_compactions();
     };
 }
 
@@ -756,12 +758,18 @@ future<> compaction_manager::postponed_compactions_reevaluation() {
             _postponed.clear();
             co_return;
         }
-        auto postponed = std::move(_postponed);
+        auto postponed = std::exchange(_postponed, {});
         try {
-            for (auto& t : postponed) {
+            while (!postponed.empty()) {
+              auto it = postponed.begin();
+              auto* t = *it;
+              postponed.erase(it);
+              if (_compaction_state.contains(t)) {
                 auto s = t->schema();
                 cmlog.debug("resubmitting postponed compaction for table {}.{} [{}]", s->ks_name(), s->cf_name(), fmt::ptr(t));
                 submit(*t);
+                co_await coroutine::maybe_yield();
+              }
             }
         } catch (...) {
             _postponed = std::move(postponed);
```

Reply:
plus the comment we talked about, explaining why the exchange is needed :)